core/src/main/java/org/apache/lucene/analysis/miscellaneous/DeDuplicatingTokenFilter.java

```diff
@@ -0,0 +1,198 @@
+
+    public DeDuplicatingTokenFilter(TokenStream in, DuplicateByteSequenceSpotter byteStreamDuplicateSpotter) {
+        this(in, byteStreamDuplicateSpotter, false);
+    }
+
+    /**
+     * 
+     * @param in
+     *            The input token stream
+     * @param byteStreamDuplicateSpotter
+     *            object which retains trie of token sequences
+     * @param emitDuplicates
+     *            true if duplicate tokens are to be emitted (use
+     *            {@link DuplicateSequenceAttribute} attribute to inspect number
+     *            of prior sightings of tokens as part of a sequence).
+     */
+    public DeDuplicatingTokenFilter(TokenStream in, DuplicateByteSequenceSpotter byteStreamDuplicateSpotter, boolean emitDuplicates) {
+        super(new DuplicateTaggingFilter(byteStreamDuplicateSpotter, in));
+        this.emitDuplicates = emitDuplicates;
+    }
+
+    @Override
+    protected boolean accept() throws IOException {
+        return emitDuplicates || seqAtt.getNumPriorUsesInASequence() < 1;
+    }
+
+    private static class DuplicateTaggingFilter extends TokenFilter {
+        private final DuplicateSequenceAttribute seqAtt = addAttribute(DuplicateSequenceAttribute.class);
+
+        TermToBytesRefAttribute termBytesAtt = addAttribute(TermToBytesRefAttribute.class);
+        private DuplicateByteSequenceSpotter byteStreamDuplicateSpotter;
+
+        protected DuplicateTaggingFilter(DuplicateByteSequenceSpotter byteStreamDuplicateSpotter, TokenStream input) {
+            super(input);
+            this.byteStreamDuplicateSpotter = byteStreamDuplicateSpotter;
+            this.windowSize = byteStreamDuplicateSpotter.getDuplicateSequenceSize();
+        }
+
+        private ArrayList<State> allTokens;
+        int pos = 0;
+        private final int windowSize;
+
+        @Override
+        public final boolean incrementToken() throws IOException {
+            if (allTokens == null) {
+                loadTokensBuffer();
+            }
+            clearAttributes();
+            if (pos < allTokens.size()) {
+                State earlierToken = allTokens.get(pos);
+                pos++;
+                restoreState(earlierToken);
+                return true;
+            } else {
+                return false;
+            }
+        }
+
+        public void loadTokensBuffer() throws IOException {
+            // Given the bytes 123456123456 and a duplicate sequence size of 6
+            // the
+            // the byteStreamDuplicateSpotter will only flag the final byte as
+            // part
+            // of a duplicate sequence due to the byte-at-a-time streaming
+            // nature of
+            // its assessments.
+            // When this happens we retain a buffer of the last 6 tokens so that
+            // we can mark
+            // the states of prior tokens (bytes 7 to 11) as also being
+            // duplicates
+            allTokens = new ArrayList<State>(256);
```

Comment:
Does that doc say that the trie uses about 6x the average document length? I wonder how much space the arraylist takes up. Probably about as much as the document length....